File: 内核源码/forward.go
package main

import (
    "io"
    "net"
    "os"
)

func forward(src, dst net.Conn) {
    io.Copy(dst, src)
    src.Close()
    dst.Close()
}

func main() {
    if len(os.Args) != 5 {
        os.Exit(1)
    }
    listenAddr := net.JoinHostPort(os.Args[1], os.Args[2])
    targetAddr := net.JoinHostPort(os.Args[3], os.Args[4])

    ln, err := net.Listen("tcp", listenAddr)
    if err != nil {
        os.Exit(1)
    }
    for {
        client, err := ln.Accept()
        if err != nil {
            continue
        }
        target, err := net.Dial("tcp", targetAddr)
        if err != nil {
            client.Close()
            continue
        }
        go forward(client, target)
        go forward(target, client)
    }
}

File: 内核源码/forward.rs
use std::net::{TcpListener, TcpStream};
use std::thread;
use std::env;

fn forward(mut src: TcpStream, mut dst: TcpStream) {
    let _ = std::io::copy(&mut src, &mut dst);
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 5 { return; }
    let listen_addr = format!("{}:{}", args[1], args[2]);
    let target_addr = format!("{}:{}", args[3], args[4]);

    let listener = TcpListener::bind(&listen_addr).unwrap();
    for stream in listener.incoming() {
        if let Ok(client) = stream {
            let addr = target_addr.clone();
            thread::spawn(move || {
                if let Ok(target) = TcpStream::connect(&addr) {
                    let c = client.try_clone().unwrap();
                    let t = target.try_clone().unwrap();
                    thread::spawn(|| forward(c, t));
                    forward(client, target);
                }
            });
        }
    }
}